### DISCUSSION TITLE
Show/hide elements depending on an external select

### DISCUSSION THREAD

**Discussion by r-chrzan** [WordPress/gutenberg#52840] [Interactivity API]
Are there plans to add a data-wp-if directive to conditionally display items? If I could help in any way, I'd be happy to, I'm testing it right now and see a few things to improve

**Comment by luisherranz** [comment]
Would it be different than [wp-show](https://github.com/WordPress/block-interactivity-experiments/issues/159)?

**Reply by r-chrzan** [reply]
Yes. The goal is to allow manage the context below for example from select. Quick example: 

I have table with many items, and each of them has an context. Now i can't change a context from select which is above the table. Maybe it would be resolved using by wp-each, or maybe it is another way to do that.

**Reply by luisherranz** [reply]
Can you share a code snippet to get a better idea? 🙂

**Reply by r-chrzan** [reply]
Sure,
I'm trying to create a table where select will manage the state of <tr> elements. Unfortunately, I can't use the context in the div above the <tr> e.g. <tbody> to create the context because the <tr> elements are generated in a loop and I have no way of getting this information beforehand.

So I'm giving myself using context all the item data I need in this case the available states (like Alaska etc.)
and I'm using a piece of code I found in this repo to update the context. All of this works on data-wp-init.

The problem occurs with the data-wp-on--change action when select does not have access to the context of the given <tr>

```
<tr
data-wp-init="actions.table.onLoad"
data-wp-context='{ "table": { "id":{{ $resource_id }}", isRestricted": false, "availableStates":<?php echo json_encode($available_states); ?>} }'
data-is-restricted="table.isRestricted"
data-value="true"
...
//rest of code
```

```
...
actions: {
       table: {
	onLoad: (store, event) => {
	const { state, selectors, context, ref } = store
	const name = ref.getAttribute('data-is-restricted')
	const value = ref.getAttribute('data-value')
	if (selectors.table.isHidden(store)) {
		const statesArray = context.table.availableStates;
		for (let i = 0; i < statesArray.length; i++) {
			if (statesArray[i].abbreviation === state.table.currentRegionCode) {
				const [key, ...path] = name.split('.').reverse();
				const obj = path.reduceRight((o, k) => o[k], context);
				obj[key] = value;
			}
		}
	}
}
...
//rest of code
```
Structure: 
```
<select
    data-wp-on--change="actions.table.onChange"
    class="select"
    >
    <option value="" >Select state</option>
    //rest options are generated by function

<table>
    <tbody>
         //loop of tr
         <tr> </tr>
        ...
        ...
        ...
    </tbody>
</table>
```

**Reply by luisherranz** [reply]
> the elements are generated in a loop and I have no way of getting this information beforehand

Hmmm… where are those elements generated? Can't you loop that data twice (once for the context, another to generate the HTML)?

I'd like to understand the use case better to see what pattern could be applied here 🙂

**Reply by r-chrzan** [reply]
Unfortunately, I can't do it earlier because I save only the first 10 results to the table. I download the rest via rest API and loop on tr elements. I know it may not seem optimal, but i need a fresh data let’s say, because the block is pulling data from ACF.

**Reply by r-chrzan** [reply]
Ok, i moved this at the top of my block and now i can generated twice, but there is another problem, even if i made loop for context i don't know how to manage state in selectors on view.js 
This is log from:
![image](https://github.com/WordPress/block-interactivity-experiments/assets/26217683/beeefbbe-a509-42c5-b490-64bcafd5aaaf)
and this is from my view.js 
<img width="395" alt="image" src="https://github.com/WordPress/block-interactivity-experiments/assets/26217683/57ce93e7-e8e0-431e-80ec-7f64fdeee356">
can i make a loop here? because `context.t2p.available` is undefined right now

**Reply by luisherranz** [reply]
I think adding a unique ID to the context and then having a map in the state to store the data is a good solution. [I actually did the same for a workshop](https://github.com/luisherranz/wceu2023/blob/main/src/blocks/quiz/render.php).

> The problem occurs with the data-wp-on--change action when select does not have access to the context of the given

Does it have access to the selected item ID?

**Reply by luisherranz** [reply]
> can i make a loop here? because context.t2p.available is undefined right now

Can you share the full code somehow? It's very difficult to follow up with so little context…

**Reply by r-chrzan** [reply]
@luisherranz i prepared repo and posted some comments to be more clearly: https://github.com/r-chrzan/interactivity-table

thanks for your help btw!

**Reply by luisherranz** [reply]
I still don't quite understand all that is going on in your code, but I made a video showing how to show/hide the different table rows depending on the value of the select:

<div>
    <a href="https://www.loom.com/share/610a1fcbb9d241cd8df98333156fe51e">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/610a1fcbb9d241cd8df98333156fe51e-with-play.gif">
    </a>
  </div>


https://www.loom.com/share/610a1fcbb9d241cd8df98333156fe51e

[I also made a pull request with the code](https://github.com/r-chrzan/interactivity-table/pull/1).

Let me know if that helps.

**Reply by r-chrzan** [reply]
@luisherranz I know, my code may be messy but I tried different ways :) Generally this is what I'm trying to achieve, which is to hide elements in the table. Maybe not necessarily all the <tr> elements that are inside the <tr>. In general, my goal in the project is to introduce geotargeting to better serve content to users. Your PR will definitely make my job easier because that's what I've been trying to do. Now I can see that I need to change my thinking a bit when it comes to Interactivity, anyway thank you very much for your help!

**Reply by luisherranz** [reply]
Awesome. I'm glad that was what you were trying to accomplish 🙂

> Maybe not necessarily all the elements that are inside

In that case, you can move or add more `data-wp-bind--hidden` directives to the different parts of the HTML to control what needs to be shown at each moment. And you can add other state and selectors if there are more conditions that should hide or show elements.

Please, don't hesitate to open a new discussion if you get stuck again.
